src/c_api.rs

```diff
@@ -0,0 +1,612 @@
+pub struct RedisModuleCtx {
+    _unused: [u8; 0],
+}
+
+#[repr(C)]
+pub enum JSONType {
+    String = 0,
+    Int = 1,
+    Double = 2,
+    Bool = 3,
+    Object = 4,
+    Array = 5,
+    Null = 6,
+}
+
+#[repr(C)]
+#[allow(non_camel_case_types)]
+pub enum RedisReturnCode {
+    REDISMODULE_OK = 0,
+    REDISMODULE_ERR = 1,
+}
+
+//---------------------------------------------------------------------------------------------
+
+pub struct JSONApiKey<'a> {
+    _key: RedisKeyWritable,
+    redis_json: &'a mut RedisJSON,
+    cstr_val: Option<CString>,
+    ctx: Context,
+}
+
+impl<'a> JSONApiKey<'a> {
+    pub fn create_rmstring(
+        &self,
+        from_str: &str,
+        str: *mut *mut rawmod::RedisModuleString,
+    ) -> c_int {
+        if let Ok(s) = CString::new(from_str) {
+            let p = s.as_bytes_with_nul().as_ptr() as *const c_char;
+            let len = s.as_bytes().len();
+            unsafe { *str = rawmod::RedisModule_CreateString.unwrap()(self.ctx.get_raw(), p, len) };
+            return RedisReturnCode::REDISMODULE_OK as c_int;
+        }
+        return RedisReturnCode::REDISMODULE_ERR as c_int;
+    }
+
+    fn get_value(&self, path: *const c_char) -> Option<&Value> {
+        let path = unsafe { CStr::from_ptr(path).to_str().unwrap() };
+        if let Ok(ref value) = self.redis_json.get_first(path) {
+            Some(value)
+        } else {
+            None
+        }
+    }
+}
+pub type JSONApiKeyRef<'a> = *mut JSONApiKey<'a>;
+
+impl<'a> JSONApiKey<'a> {
+    pub fn new_from_key(key: RedisKeyWritable, ctx: Context) -> Result<JSONApiKey<'a>, RedisError> {
+        let res = key.get_value::<RedisJSON>(&REDIS_JSON_TYPE)?;
+        if let Some(value) = res {
+            Ok(JSONApiKey {
+                _key: key,
+                redis_json: value,
+                cstr_val: None,
+                ctx: ctx,
+            })
+        } else {
+            Err(RedisError::Str("Not a JSON key"))
+        }
+    }
+
+    pub fn new_from_redis_string(
+        ctx: *mut rawmod::RedisModuleCtx,
+        key_str: *mut rawmod::RedisModuleString,
+    ) -> Result<JSONApiKey<'a>, RedisError> {
+        let ctx = Context::new(ctx);
+        let key = ctx.open_with_redis_string(key_str);
+        JSONApiKey::new_from_key(key, ctx)
+    }
+
+    pub fn new_from_str(
+        ctx: *mut rawmod::RedisModuleCtx,
+        path: *const c_char,
+    ) -> Result<JSONApiKey<'a>, RedisError> {
+        let ctx = Context::new(ctx);
+        let path = unsafe { CStr::from_ptr(path).to_str().unwrap() };
+        let key = ctx.open_key_writable(path);
+        JSONApiKey::new_from_key(key, ctx)
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn JSONAPI_openKey<'a>(
+    ctx: *mut rawmod::RedisModuleCtx,
+    key_str: *mut rawmod::RedisModuleString,
+) -> JSONApiKeyRef<'a> {
+    match JSONApiKey::new_from_redis_string(ctx, key_str) {
+        Ok(key) => Box::into_raw(Box::new(key)) as JSONApiKeyRef,
+        _ => null_mut(),
+    }
+}
```

Comment:
It might be better to use `bindgen` (as in `redismodule-rs`) to create function stubs from the header file. Less code to maintain, easier to change in the future and less chances of errors.

Reply:
Agree, but I suggest to switch to `bindgen` in the next "phase" (first stabilize and merge to master)